src/platform_impl/linux/drm/event_loop.rs

```diff
@@ -0,0 +1,1417 @@
+        pointer::PointerScrollEvent,
+        tablet_pad::{ButtonState, KeyState},
+        tablet_tool::{TabletToolEventTrait, TipState},
+        touch::{TouchEventPosition, TouchEventSlot},
+    },
+    LibinputInterface,
+};
+
+use crate::{
+    dpi::PhysicalPosition,
+    event::{ElementState, Force, KeyboardInput, ModifiersState, MouseScrollDelta, StartCause},
+    event_loop::{ControlFlow, EventLoopClosed},
+    platform::unix::Card,
+    platform_impl::{platform::sticky_exit_callback, xkb_keymap},
+};
+
+struct Interface(libseat::Seat, HashMap<RawFd, i32>);
+
+impl LibinputInterface for Interface {
+    fn open_restricted(&mut self, path: &Path, _flags: i32) -> Result<RawFd, i32> {
+        self.0
+            .open_device(&path)
+            .map(|(id, file)| {
+                self.1.insert(file, id);
+                file
+            })
+            .map_err(|err| err.into())
+    }
+    fn close_restricted(&mut self, fd: RawFd) {
+        if let Some(dev) = self.1.get(&fd).copied() {
+            self.0.close_device(dev).unwrap();
+        }
+        unsafe { std::fs::File::from_raw_fd(fd) };
+    }
+}
+
+pub struct LibinputInputBackend {
+    context: input::Libinput,
+    xkb_ctx: xkb::State,
+    xkb_keymap: xkb::Keymap,
+    xkb_compose: xkb::compose::State,
+    token: Token,
+    touch_location: PhysicalPosition<f64>,
+    screen_size: (u32, u32),
+    modifiers: ModifiersState,
+    cursor_positon: PhysicalPosition<f64>,
+    timer_handle: calloop::timer::TimerHandle<(KeyboardInput, Option<char>)>,
+    // cursor_plane: drm::control::plane::Handle,
+    // cursor_buffer: drm::control::framebuffer::Handle,
+}
+
+impl LibinputInputBackend {
+    /// Initialize a new [`LibinputInputBackend`] from a given already initialized
+    /// [libinput context](libinput::Libinput).
+    pub fn new(
+        context: input::Libinput,
+        screen_size: (u32, u32),
+        timer_handle: calloop::timer::TimerHandle<(KeyboardInput, Option<char>)>,
+        xkb_ctx: xkb::State,
+        xkb_keymap: xkb::Keymap,
+        xkb_compose: xkb::compose::State,
+        // cursor_plane: drm::control::plane::Handle,
+        // cursor_buffer: drm::control::framebuffer::Handle
+    ) -> Self {
+        LibinputInputBackend {
+            context,
+            token: Token::invalid(),
+            touch_location: PhysicalPosition::new(0.0, 0.0),
+            cursor_positon: PhysicalPosition::new(0.0, 0.0),
+            modifiers: ModifiersState::empty(),
+            screen_size,
+            timer_handle,
+            xkb_ctx,
+            xkb_keymap,
+            xkb_compose,
+            // cursor_buffer,
+            // cursor_plane,
+        }
+    }
+}
+
+impl AsRawFd for LibinputInputBackend {
+    fn as_raw_fd(&self) -> RawFd {
+        self.context.as_raw_fd()
+    }
+}
+
+impl EventSource for LibinputInputBackend {
+    type Event = crate::event::Event<'static, ()>;
+    type Metadata = ();
+    type Ret = ();
+
+    fn process_events<F>(
```

Comment:
This function is barely readable due to overly complex nesting. Certain names are just one character. First of all I'd suggest to extract event handling parts.

reworked object creation, since you're using full pathes all over the place, could try looking into other backends which try aliasing backend stuff with `PlatformDeviceId` and crate public API left as is e.g. `DeviceId`.

This alone should reduce lots of code here and make it more readable.

Reply:
Fixed